lib/tom_queue/deferred_work_manager.rb

```diff
@@ -113,6 +121,24 @@ def stop
       deferred_set && deferred_set.interrupt
     end
 
-  end
+    # Register handlers to shut down the manager gracefully
+    def register_signal_handlers
+      Thread.main[:signals] = []
+
+      [:QUIT, :INT, :TERM].each do |sig|
+        trap(sig) do
+          Thread.main[:signals] << sig
```

Comment:
This process isn't doing any work, it's just shunting messages around  - it's purely a timeout on the blocking "pop" operation. I've just had a quick chat with @cfabianski and I think his approach is cleaner, we just need to not try to grab the mutex in the trap context to avoid the `ThreadError`.